### PR TITLE
Clean up code for nonrelative path completions

### LINKED ISSUE #23096: pathCompletions should use getEmitModuleResolutionKind
Inside `pathCompletions`, we check `compilerOptions.moduleResolution` - we should be using `getEmitModuleResolutionKind` to check instead so that defaults based on module target are upheld. (In effect, this means we're not offering `node_modules` completion in places where we should right now)

### PR DESCRIPTION
**Pull request by ghost**
Fixes #23096 

* Only check `moduleResolutionKind` in one place. And correctly check it using `getEmitModuleResolutionKind`.
* This uncovered duplicate completions when we add completions from `package.json` *and* add completions from searching through `node_modules`. Now, we will only do the latter if the former failed.
* Searching through ambient modules was moved to before this as it doesn't depend on `moduleResolutionKind`.

### CODE REVIEW

File: tests/cases/fourslash/completionListInImportClause05.ts [hunk 1]
```diff
@@ -12,4 +12,4 @@
 // NOTE: The node_modules folder is in "/", rather than ".", because it requires
 // less scaffolding to mock.  In particular, "/" is where we look for type roots.
 
-verify.completionsAt("1", ["@a/b", "@c/d", "@e/f"], { isNewIdentifierLocation: true });
+verify.completionsAt("1", ["@e/f", "@a/b", "@c/d"], { isNewIdentifierLocation: true });
```
[the remark this answers is not in the document]
Comment (weswigham) [reply tests/cases/fourslash/completionListInImportClause05.ts]: Shouldn't the results be sorted?

File: src/services/pathCompletions.ts [hunk 2]
```diff
@@ -262,45 +275,31 @@ namespace ts.Completions.PathCompletions {
...
-                        for (let f of nestedFiles) {
-                            f = normalizePath(f);
-                            const nestedModule = removeFileExtension(getBaseFileName(f));
-                            nonRelativeModuleNames.push(nestedModule);
-                        }
-                    }
-                }
+    function getNamesFromVisibleNodeModules(fragmentDirectory: string | undefined, scriptPath: string, host: LanguageServiceHost): string[] {
+        return flatMap(enumerateNodeModulesVisibleToScript(host, scriptPath), visibleModule => {
+            if (fragmentDirectory === undefined) {
```
Comment (sheetalkamat) [inline src/services/pathCompletions.ts]: Isnt this always undefined since its called only when fragmentDirectory ===undefined

File: tests/cases/fourslash/completionListInImportClause05.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (andy-ms) [reply tests/cases/fourslash/completionListInImportClause05.ts]: They weren't sorted by name before, except by coincidence. We could make a PR sorting them but it would change a bunch of baselines. I think editors do their own sorting on the result anyway.

File: src/services/pathCompletions.ts (hunk 2 above)
Comment (andy-ms) [reply src/services/pathCompletions.ts]: Wow, if we're always in the first case we can clean out a lot of code.

File: tests/cases/fourslash/completionListInImportClause05.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (weswigham) [reply tests/cases/fourslash/completionListInImportClause05.ts]: Alright.

File: src/services/pathCompletions.ts [hunk 3]
```diff
@@ -177,19 +176,33 @@ namespace ts.Completions.PathCompletions {
...
+                    if (!result.some(entry => entry.name === moduleName)) {
+                        result.push(nameAndKind(moduleName, ScriptElementKind.externalModuleName));
+                    }
+                }
+                foundGlobal = result.length !== oldLength;
+            }
+            if (!foundGlobal) {
+                forEachAncestorDirectory(scriptPath, ancestor => {
+                    const nodeModules = combinePaths(ancestor, "node_modules");
+                    if (host.directoryExists(nodeModules)) {
```
Comment (weswigham) [inline src/services/pathCompletions.ts]: `host.directoryExists` is optional, right?

File: src/compiler/moduleNameResolver.ts [hunk 4]
```diff
@@ -117,9 +117,10 @@ namespace ts {
         }
     }
 
-    function readJson(path: string, host: ModuleResolutionHost): PackageJson {
+    /* @internal */
+    export function readJson(path: string, host: { readFile(fileName: string): string | undefined }): object {
```
Comment (weswigham) [inline src/compiler/moduleNameResolver.ts]: Why not mark `readFile` optional?

Comment (andy-ms) [reply src/compiler/moduleNameResolver.ts]: Whoops, left in a change to the function body on accident. This should not be optional here because it is checked by the callers.

File: src/services/pathCompletions.ts [hunk 5]
```diff
@@ -177,19 +176,33 @@ namespace ts.Completions.PathCompletions {
...
-
-        for (const moduleName of enumeratePotentialNonRelativeModules(fragment, scriptPath, compilerOptions, typeChecker, host)) {
-            result.push(nameAndKind(moduleName, ScriptElementKind.externalModuleName));
+        if (getEmitModuleResolutionKind(compilerOptions) === ModuleResolutionKind.NodeJs) {
+            // If looking for a global package name, don't just include everything in `node_modules` because that includes dependencies' own dependencies.
+            // (But do if we didn't find anything, e.g. 'package.json' missing.)
+            let foundGlobal = false;
+            if (fragmentDirectory === undefined) {
+                const oldLength = result.length;
+                getCompletionEntriesFromTypings(host, compilerOptions, scriptPath, result);
```
Comment (sheetalkamat) [inline src/services/pathCompletions.ts]: earlier this was done irrespective of fragment/module emit .. Why the change?

Comment (andy-ms) [reply src/services/pathCompletions.ts]: :+1: good catch

**Comment by andy-ms** [comment]
@weswigham Good to go?
